test/utils/oidc/testserver.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oidc
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/tls"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	openIDWellKnownWebPath = "/.well-known/openid-configuration"
+	authWebPath            = "/auth"
+	tokenWebPath           = "/token"
+	jwksWebPath            = "/jwks"
+	userInfoWebPath        = "/user-info"
+)
+
+var (
+	ErrRefreshTokenExpired   = errors.New("refresh token is expired")
+	ErrClientSecretNotPassed = errors.New("client secret is not passed")
+	ErrBadClientSecret       = errors.New("client secret is bad")
+	ErrBadClientID           = errors.New("client ID is bad")
+)
+
+type TestServer struct {
+	httpServer   *httptest.Server
+	tokenHandler *MockTokenHandler
+	jwksHandler  *MockJWKsHandler
+	TearDownFn   func()
```

Comment:
Avoid tear down functions.  Use `t.Cleanup` inside of functions like `BuildAndRunTestServer` to match sure that tear happens automatically.

Reply:
Just borrowed the `TearDownFn` idea from the [testing API Server](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/testing/testserver.go) :) Ok, updated according to your proposal